### ISSUE TITLE
kub-aws 0.10.2/k8s 1.9: Unable to reconfigure etcd member

### ISSUE THREAD

**Issue by flah00** [kubernetes-incubator/kube-aws#1478]
Our etcd cluster seems to have lost its ability to replace a node. From my reading of the logs, it seems that one of the node's hashed ids is not being computed correctly. Any thoughts on how we can help our etcd cluster recover?

We ended up in this spot because one of the nodes suddenly ran out of CPU credits. We terminated the instance, and since then, it has been unable to re-join the cluster. As the `ETCD0` node attempts to rejoin, the logs seem to suggest that its hash for `ETCD2` is inconsistently computed.

## cluster state

```console
ETCD0 core # set -a ; source /etc/etcd-environment ; source /var/run/coreos/etcdadm-environment ; set +a
ETCD0 core # /opt/bin/etcdadm member_etcdctl member list
5d900f13cc207b88, started, etcd2, https://ETCD2.compute-1.amazonaws.com:2380, https://ETCD2.compute-1.amazonaws.com:2379
931b8e263ac72b95, started, etcd1, https://ETCD1.compute-1.amazonaws.com:2380, https://ETCD1.compute-1.amazonaws.com:2379

ETCD0 core # /opt/bin/etcdadm check
https://ETCD0.compute-1.amazonaws.com:2379 is unhealthy: failed to connect: dial tcp ETCD0:2379: getsockopt: connection refused
Error:  unhealthy cluster
https://ETCD0.compute-1.amazonaws.com:2379 is unhealthy: failed to connect: dial tcp ETCD0:2379: getsockopt: connection refused
Error:  unhealthy cluster
https://ETCD1.compute-1.amazonaws.com:2379 is healthy: successfully committed proposal: took = 1.890425ms
https://ETCD2.compute-1.amazonaws.com:2379 is healthy: successfully committed proposal: took = 2.519788ms
/opt/bin/etcdadm: info: cluster_is_unhealthy: quorum=2 healthy=2
/opt/bin/etcdadm: info: cluster_is_unhealthy: cluster is healthy
```

## recovery attempt

I terminated the `ETCD0` instance, in the hopes that a clean recovery attempt would fix the issues. I found that the snapshot on `ETCD0` is only 17M, while the snapshot db on `ETCD1` and `ETCD2` is 138M, and the s3 snapshot is 130M. I then tried to follow the instructions 

### etcdadm replace output

On the newly replaced instance, I tried to follow the manual recovery steps [here](https://kubernetes-incubator.github.io/kube-aws/advanced-topics/etcd-backup-and-restore.html). The command does not return successfully.

```console
ETCD0 core # set -a; source /var/run/coreos/etcdadm-environment; set +a
ETCD0 core # /opt/bin/etcdadm replace
/opt/bin/etcdadm: line 81: ETCD_INITIAL_CLUSTER: unbound variable
/opt/bin/etcdadm: line 761: peers[$1]: unbound variable
/opt/bin/etcdadm: line 81: ETCD_INITIAL_CLUSTER: unbound variable
/opt/bin/etcdadm: line 769: peers[$1]: unbound variable
/opt/bin/etcdadm: line 81: ETCD_INITIAL_CLUSTER: unbound variable
/opt/bin/etcdadm: line 761: peers[$1]: unbound variable
/opt/bin/etcdadm: line 81: ETCD_INITIAL_CLUSTER: unbound variable
/opt/bin/etcdadm: line 761: peers[$1]: unbound variable
/opt/bin/etcdadm: info: member_clean_data_dir: cleaning data dir of
/opt/bin/etcdadm: info: member_clean_data_dir: data dir /home/core/work/ does not exist. nothing to remove
/opt/bin/etcdadm: info: member_replace_failed: connecting to https://ETCD1.compute-1.amazonaws.com:2379
5d900f13cc207b88: name=etcd2 peerURLs=https://ETCD2.compute-1.amazonaws.com:2380 clientURLs=https://ETCD2.compute-1.amazonaws.com:2379 isLeader=false
931b8e263ac72b95: name=etcd1 peerURLs=https://ETCD1.compute-1.amazonaws.com:2380 clientURLs=https://ETCD1.compute-1.amazonaws.com:2379 isLeader=true
/opt/bin/etcdadm: info: member_replace_failed: removing member 5d900f13cc207b88
931b8e263ac72b95
Couldn't find a member in the cluster with an ID of 5d900f13cc207b88
931b8e263ac72b95.
ETCD0 core # echo $?
1
```

## etcd0 etcd-member logs

The `etcdadm member_etcdctl member list` command shows the `ETCD2` hash as `5d900f13cc207b88`. This does not match the hash in `ETCD0` logs, which lists `ETCD2` hash as  `9d1da0da9a0193ba`.

```
Nov 01 18:21:44 ETCD0.ec2.internal systemd[1]: Starting etcd (System Application Container)...
Nov 01 18:21:44 ETCD0.ec2.internal systemctl[2236]: active
Nov 01 18:21:44 ETCD0.ec2.internal etcd-wrapper[2240]: ++ id -u etcd
Nov 01 18:21:44 ETCD0.ec2.internal etcd-wrapper[2240]: + exec /usr/bin/rkt run --uuid-file-save=/var/lib/coreos/etcd-member-wrapper.uuid --trust-keys-from-https --mount volume=coreos-systemd-dir,target=/run/systemd/system --volume coreos-systemd-dir,kind=host,source=/run/systemd/system,readOnly=true --mount volume=coreos-notify,target=/run/systemd/notify --volume coreos-notify,kind=host,source=/run/systemd/notify --set-env=NOTIFY_SOCKET=/run/systemd/notify --volume coreos-data-dir,kind=host,source=/var/lib/etcd2,readOnly=false --volume coreos-etc-ssl-certs,kind=host,source=/etc/ssl/certs,readOnly=true --volume coreos-usr-share-certs,kind=host,source=/usr/share/ca-certificates,readOnly=true --volume coreos-etc-hosts,kind=host,source=/etc/hosts,readOnly=true --volume coreos-etc-resolv,kind=host,source=/etc/resolv.conf,readOnly=true --mount volume=coreos-data-dir,target=/var/lib/etcd --mount volume=coreos-etc-ssl-certs,target=/etc/ssl/certs --mount volume=coreos-usr-share-certs,target=/usr/share/ca-certificates --mount volume=coreos-etc-hosts,target=/etc/hosts --mount volume=coreos-etc-resolv,target=/etc/resolv.conf --inherit-env --stage1-from-dir=stage1-fly.aci quay.io/coreos/etcd:v3.2.7 --user=232 --
Nov 01 18:21:49 ETCD0.ec2.internal etcd-wrapper[2240]: pubkey: prefix: "quay.io/coreos/etcd"
Nov 01 18:21:49 ETCD0.ec2.internal etcd-wrapper[2240]: key: "https://quay.io/aci-signing-key"
Nov 01 18:21:49 ETCD0.ec2.internal etcd-wrapper[2240]: gpg key fingerprint is: BFF3 13CD AA56 0B16 A898  7B8F 72AB F5F6 799D 33BC
Nov 01 18:21:49 ETCD0.ec2.internal etcd-wrapper[2240]:         Quay.io ACI Converter (ACI conversion signing key) <support@quay.io>
Nov 01 18:21:49 ETCD0.ec2.internal etcd-wrapper[2240]: Trusting "https://quay.io/aci-signing-key" for prefix "quay.io/coreos/etcd" without fingerprint review.
Nov 01 18:21:49 ETCD0.ec2.internal etcd-wrapper[2240]: Added key for prefix "quay.io/coreos/etcd" at "/etc/rkt/trustedkeys/prefix.d/quay.io/coreos/etcd/bff313cdaa560b16a8987b8f72abf5f6799d33bc"
Nov 01 18:21:49 ETCD0.ec2.internal etcd-wrapper[2240]: Downloading signature:  0 B/473 B
Nov 01 18:21:49 ETCD0.ec2.internal etcd-wrapper[2240]: Downloading signature:  473 B/473 B
Nov 01 18:21:49 ETCD0.ec2.internal etcd-wrapper[2240]: Downloading signature:  473 B/473 B
Nov 01 18:21:50 ETCD0.ec2.internal etcd-wrapper[2240]: Downloading ACI:  0 B/13.3 MB
Nov 01 18:21:50 ETCD0.ec2.internal etcd-wrapper[2240]: Downloading ACI:  16.4 KB/13.3 MB
Nov 01 18:21:51 ETCD0.ec2.internal etcd-wrapper[2240]: Downloading ACI:  13.3 MB/13.3 MB
Nov 01 18:21:51 ETCD0.ec2.internal etcd-wrapper[2240]: image: signature verified:
Nov 01 18:21:51 ETCD0.ec2.internal etcd-wrapper[2240]:   Quay.io ACI Converter (ACI conversion signing key) <support@quay.io>
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.580972 I | pkg/flags: recognized and used environment variable ETCD_ADVERTISE_CLIENT_URLS=https://ETCD0.compute-1.amazonaws.com:2379
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.581039 I | pkg/flags: recognized and used environment variable ETCD_CERT_FILE=/etc/ssl/certs/etcd.pem
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.581046 I | pkg/flags: recognized and used environment variable ETCD_CLIENT_CERT_AUTH=true
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.581056 I | pkg/flags: recognized and used environment variable ETCD_DATA_DIR=/var/lib/etcd
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.581076 I | pkg/flags: recognized and used environment variable ETCD_INITIAL_ADVERTISE_PEER_URLS=https://ETCD0.compute-1.amazonaws.com:2380
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.581086 I | pkg/flags: recognized and used environment variable ETCD_INITIAL_CLUSTER=etcd0=https://ETCD0.compute-1.amazonaws.com:2380,etcd1=https://ETCD1.compute-1.amazonaws.com:2380,etcd2=https://ETCD2.compute-1.amazonaws.com:2380
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.581093 I | pkg/flags: recognized and used environment variable ETCD_INITIAL_CLUSTER_STATE=new
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.581100 I | pkg/flags: recognized and used environment variable ETCD_KEY_FILE=/etc/ssl/certs/etcd-key.pem
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.581112 I | pkg/flags: recognized and used environment variable ETCD_LISTEN_CLIENT_URLS=https://0.0.0.0:2379
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.581119 I | pkg/flags: recognized and used environment variable ETCD_LISTEN_PEER_URLS=https://ETCD0:2380
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.581129 I | pkg/flags: recognized and used environment variable ETCD_NAME=etcd0
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.581136 I | pkg/flags: recognized and used environment variable ETCD_PEER_CERT_FILE=/etc/ssl/certs/etcd.pem
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.581144 I | pkg/flags: recognized and used environment variable ETCD_PEER_KEY_FILE=/etc/ssl/certs/etcd-key.pem
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.581152 I | pkg/flags: recognized and used environment variable ETCD_PEER_TRUSTED_CA_FILE=/etc/ssl/certs/etcd-trusted-ca.pem
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.581168 I | pkg/flags: recognized and used environment variable ETCD_TRUSTED_CA_FILE=/etc/ssl/certs/etcd-trusted-ca.pem
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.581182 W | pkg/flags: unrecognized environment variable ETCD_USER=etcd
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.581187 W | pkg/flags: unrecognized environment variable ETCD_IMAGE_TAG=v3.2.7
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.581215 I | etcdmain: etcd Version: 3.2.7
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.581221 I | etcdmain: Git SHA: bb66589
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.581224 I | etcdmain: Go Version: go1.8.3
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.581229 I | etcdmain: Go OS/Arch: linux/amd64
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.581233 I | etcdmain: setting maximum number of CPUs to 2, total number of available CPUs is 2
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.581289 I | embed: peerTLS: cert = /etc/ssl/certs/etcd.pem, key = /etc/ssl/certs/etcd-key.pem, ca = , trusted-ca = /etc/ssl/certs/etcd-trusted-ca.pem, client-cert-auth = false
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.581944 I | embed: listening for peers on https://ETCD0:2380
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.581986 I | embed: listening for client requests on 0.0.0.0:2379
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.590975 I | pkg/netutil: resolving ETCD0.compute-1.amazonaws.com:2380 to 34.233.219.195:2380
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.591368 I | pkg/netutil: resolving ETCD0.compute-1.amazonaws.com:2380 to 34.233.219.195:2380
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.599025 I | etcdserver: name = etcd0
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.599047 I | etcdserver: data dir = /var/lib/etcd
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.599055 I | etcdserver: member dir = /var/lib/etcd/member
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.599059 I | etcdserver: heartbeat = 100ms
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.599063 I | etcdserver: election = 1000ms
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.599067 I | etcdserver: snapshot count = 100000
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.599079 I | etcdserver: advertise client URLs = https://ETCD0.compute-1.amazonaws.com:2379
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.599085 I | etcdserver: initial advertise peer URLs = https://ETCD0.compute-1.amazonaws.com:2380
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.599097 I | etcdserver: initial cluster = etcd0=https://ETCD0.compute-1.amazonaws.com:2380,etcd1=https://ETCD1.compute-1.amazonaws.com:2380,etcd2=https://ETCD2.compute-1.amazonaws.com:2380
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.605094 I | etcdserver: starting member 7ff2475ad25e7c04 in cluster 94a19d5374fac8e1
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.605143 I | raft: 7ff2475ad25e7c04 became follower at term 0
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.605154 I | raft: newRaft 7ff2475ad25e7c04 [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.605159 I | raft: 7ff2475ad25e7c04 became follower at term 1
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.614127 W | auth: simple token is not cryptographically signed
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.618755 I | rafthttp: starting peer 931b8e263ac72b95...
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.618796 I | rafthttp: started HTTP pipelining with peer 931b8e263ac72b95
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.620844 I | rafthttp: started streaming with peer 931b8e263ac72b95 (writer)
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.620886 I | rafthttp: started streaming with peer 931b8e263ac72b95 (writer)
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.621823 I | rafthttp: started peer 931b8e263ac72b95
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.621974 I | rafthttp: added peer 931b8e263ac72b95
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.622004 I | rafthttp: starting peer 9d1da0da9a0193ba...
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.622017 I | rafthttp: started HTTP pipelining with peer 9d1da0da9a0193ba
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.624281 I | rafthttp: started streaming with peer 931b8e263ac72b95 (stream Message reader)
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.624790 I | rafthttp: started peer 9d1da0da9a0193ba
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.624817 I | rafthttp: added peer 9d1da0da9a0193ba
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.624846 I | etcdserver: starting server... [version: 3.2.7, cluster version: to_be_decided]
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.624863 I | embed: ClientTLS: cert = /etc/ssl/certs/etcd.pem, key = /etc/ssl/certs/etcd-key.pem, ca = , trusted-ca = /etc/ssl/certs/etcd-trusted-ca.pem, client-cert-auth = true
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.625413 I | rafthttp: started streaming with peer 9d1da0da9a0193ba (writer)
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.625562 I | rafthttp: started streaming with peer 9d1da0da9a0193ba (writer)
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.625675 I | rafthttp: started streaming with peer 931b8e263ac72b95 (stream MsgApp v2 reader)
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.625888 I | rafthttp: started streaming with peer 9d1da0da9a0193ba (stream MsgApp v2 reader)
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.626089 I | rafthttp: started streaming with peer 9d1da0da9a0193ba (stream Message reader)
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.626770 I | etcdserver/membership: added member 7ff2475ad25e7c04 [https://ETCD0.compute-1.amazonaws.com:2380] to cluster 94a19d5374fac8e1
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.626872 I | etcdserver/membership: added member 931b8e263ac72b95 [https://ETCD1.compute-1.amazonaws.com:2380] to cluster 94a19d5374fac8e1
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.626930 I | etcdserver/membership: added member 9d1da0da9a0193ba [https://ETCD2.compute-1.amazonaws.com:2380] to cluster 94a19d5374fac8e1
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.635127 E | etcdserver: the member has been permanently removed from the cluster
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.635144 I | etcdserver: the data-dir used by this member must be removed.
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.635165 I | etcdserver: aborting publish because server is stopped
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.635178 I | rafthttp: stopping peer 931b8e263ac72b95...
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.635186 I | rafthttp: stopped streaming with peer 931b8e263ac72b95 (writer)
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.635193 I | rafthttp: stopped streaming with peer 931b8e263ac72b95 (writer)
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.635270 I | rafthttp: stopped HTTP pipelining with peer 931b8e263ac72b95
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.635289 I | rafthttp: stopped streaming with peer 931b8e263ac72b95 (stream MsgApp v2 reader)
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.635301 I | rafthttp: stopped streaming with peer 931b8e263ac72b95 (stream Message reader)
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.635307 I | rafthttp: stopped peer 931b8e263ac72b95
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.635311 I | rafthttp: stopping peer 9d1da0da9a0193ba...
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.635317 I | rafthttp: stopped streaming with peer 9d1da0da9a0193ba (writer)
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.635323 I | rafthttp: stopped streaming with peer 9d1da0da9a0193ba (writer)
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.635332 I | rafthttp: stopped HTTP pipelining with peer 9d1da0da9a0193ba
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.635343 I | rafthttp: stopped streaming with peer 9d1da0da9a0193ba (stream MsgApp v2 reader)
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.635353 I | rafthttp: stopped streaming with peer 9d1da0da9a0193ba (stream Message reader)
Nov 01 18:21:52 ETCD0.ec2.internal etcd-wrapper[2240]: 2018-11-01 18:21:52.635358 I | rafthttp: stopped peer 9d1da0da9a0193ba
Nov 01 18:21:52 ETCD0.ec2.internal systemd[1]: Started etcd (System Application Container).
Nov 01 18:21:52 ETCD0.ec2.internal rkt[2287]: pod "57851120-2ac2-49fa-88b4-1f9f98ce608c" is already stopped
Nov 01 18:21:52 ETCD0.ec2.internal systemd[1]: Stopped etcd (System Application Container).
```

**Comment by davidmccormick** [comment]
Hi @flah00 - that sounds painful!  Which node was it that ran out of credits?  From your `etcdadm member_etcdctl member list` listing it looks as though only ETCD0 is unhealthy - with ETCD1 and 2 fine - meaning that you still have quorum, which is good!

I think that we should be able to replace ETCD0 and get it to rejoin the cluster and I see you ran etcdadm replace to do this but it clearly had some issues: -
```
/opt/bin/etcdadm: line 81: ETCD_INITIAL_CLUSTER: unbound variable
/opt/bin/etcdadm: line 761: peers[$1]: unbound variable
/opt/bin/etcdadm: line 81: ETCD_INITIAL_CLUSTER: unbound variable
/opt/bin/etcdadm: line 769: peers[$1]: unbound variable
/opt/bin/etcdadm: line 81: ETCD_INITIAL_CLUSTER: unbound variable
/opt/bin/etcdadm: line 761: peers[$1]: unbound variable
/opt/bin/etcdadm: line 81: ETCD_INITIAL_CLUSTER: unbound variable
/opt/bin/etcdadm: line 761: peers[$1]: unbound variable
/opt/bin/etcdadm: info: member_clean_data_dir: cleaning data dir of
```

Looking at the script, it looks to be missing some required environment variables and I think that you probably also need to `source /etc/etcd-environment` as well as `/var/run/coreos/etcdadm-environment`.  Could you try that and run the `etcdadm replace` again and let me know what happens?

**Comment by davidmccormick** [comment]
Hi @flah00, this is an interesting problem! :)  I've not experienced an issue like this one... would you be so kind as to run `etcdadm check` on *each* of the etcd servers so we can compare?

My gut feeling would be to go off-script somewhat and recover the cluster as though we only have one good etcd server left (probably etcd1).  You should backup each of the etcd data directories so you can keep them safe.  Then try making etcd1 a 1 node etcd cluster starting with `--force-new-cluster` flag (more details here https://coreos.com/etcd/docs/latest/v2/admin_guide.html) before manually adding back in etcd0 and etcd2.  Let me know if you need more help with this and maybe I could write you a procedure?

**Comment by flah00** [comment]
@davidmccormick Interesting indeed! 

Thanks for the link, I'll read it over. I do have some questions, which I assume the link does not cover. I have etcd disaster recovery turned on (kube-aws cluster.yaml `etcd.disasterRecovery.automated=true`). Are there other services I should disable, before starting this adventure? Also, as I work through this, what state will the kubernetes cluster find itself in, ie will cronjobs still get scheduled, will running pods hum along? 

Finally, if the manual reconstitution of the etcd cluster fails, I assume that to revert, I would run the following on each node

1. `systemctl stop etcd-member`
1.  `find /var/lib/etcd2/member -type delete`
1. `cp db.orig /var/lib/etcd2/member/snap`
1. `systemctl start etcd-member`

Here are the results of `etcdadm check` from all of the nodes. 

```
## ETCD0

ETCD0 core # set -a; source /var/run/coreos/etcdadm-environment; source /etc/etcd-environment; source /var/run/coreos/etcdadm-environment; set +a
ETCD0 core # /opt/bin/etcdadm check
https://ETCD0.compute-1.amazonaws.com:2379 is unhealthy: failed to connect: dial tcp ETCD0:2379: getsockopt: connection refused
Error:  unhealthy cluster
https://ETCD0.compute-1.amazonaws.com:2379 is unhealthy: failed to connect: dial tcp ETCD0:2379: getsockopt: connection refused
Error:  unhealthy cluster
https://ETCD1.compute-1.amazonaws.com:2379 is healthy: successfully committed proposal: took = 2.583766ms
https://ETCD2.compute-1.amazonaws.com:2379 is healthy: successfully committed proposal: took = 1.798549ms
/opt/bin/etcdadm: info: cluster_is_unhealthy: quorum=2 healthy=2
/opt/bin/etcdadm: info: cluster_is_unhealthy: cluster is healthy

## ETCD1

ETCD1 core # set -a; source /var/run/coreos/etcdadm-environment; source /etc/etcd-environment; source /var/run/coreos/etcdadm-environment; set +a
ETCD1 core # /opt/bin/etcdadm check
https://ETCD0.compute-1.amazonaws.com:2379 is unhealthy: failed to connect: dial tcp ETCD0:2379: getsockopt: connection refused
Error:  unhealthy cluster
https://ETCD0.compute-1.amazonaws.com:2379 is unhealthy: failed to connect: dial tcp ETCD0:2379: getsockopt: connection refused
Error:  unhealthy cluster
https://ETCD1.compute-1.amazonaws.com:2379 is healthy: successfully committed proposal: took = 2.583766ms
https://ETCD2.compute-1.amazonaws.com:2379 is healthy: successfully committed proposal: took = 1.798549ms
/opt/bin/etcdadm: info: cluster_is_unhealthy: quorum=2 healthy=2

## ETCD2

ETCD2 core # set -a; source /var/run/coreos/etcdadm-environment; source /etc/etcd-environment; source /var/run/coreos/etcdadm-environment; set +a
ETCD2 core # /opt/bin/etcdadm check
https://ETCD2.compute-1.amazonaws.com:2379 is healthy: successfully committed proposal: took = 5.919529ms
https://ETCD0.compute-1.amazonaws.com:2379 is unhealthy: failed to connect: dial tcp ETCD0:2379: getsockopt: connection refused
Error:  unhealthy cluster
https://ETCD1.compute-1.amazonaws.com:2379 is healthy: successfully committed proposal: took = 2.549606ms
https://ETCD2.compute-1.amazonaws.com:2379 is healthy: successfully committed proposal: took = 1.883534ms
/opt/bin/etcdadm: info: cluster_is_unhealthy: quorum=2 healthy=2
```

Thanks for all of your help!

**Comment by davidmccormick** [comment]
Hi @flah00 - apologies for the slow reply!

Yes, you are right - there is a risk here and that is why I'm not super confident in guiding you fix your etcd by asking questions on a support ticket such as this.  I think you will have to learn the procedure yourself and tackle it with care.  If you take a backup of your data (and you can also export all of the etcd keys too for good measure) then you can always bring everything back, but you are again correct that if you encounter an issue you will have a kubernetes API outage because your etcd cluster will lose quorum.

From your etcdadm check outputs I'm happy you have a healthy cluster with only etcd0 which is unhealthy.  Personally, I would use `etcdctl` to work on the cluster rather than `etcdadm` because etcdctl is the tool I am familiar with and is what is used in the online documentation from coreos.  You can run etcdctl yourself using docker, e.g.

```
docker run --rm -e ETCDCTL_CACERT=/etc/ssl/certs/etcd-trusted-ca.pem -e ETCDCTL_CERT=/etc/ssl/certs/etcd-client.pem -e ETCDCTL_KEY=/etc/ssl/certs/etcd-client-key.pem --volume=/etc/ssl/certs:/etc/ssl/certs --env ETCDCTL_API=3 --network=host --volume=/var/run/coreos/etcdadm/snapshots:/snapshots --volume=/var/lib/etcd2:/var/lib/etcd --volume snapshots:/var/run/coreos/etcdadm/snapshots quay.io/coreos/etcd:v3.2.13 etcdctl --endpoints https://ip-10-29-25-218.us-west-2.compute.internal:2379 endpoint health
```

**Comment by arpad9** [comment]
Update @davidmccormick (and thanks for your help).  I work with Phil and I got etcd0 to join the cluster.  The causal problem is that etcdadm is trying to connect to an API endpoint via etcdctl to get various things, e.g.:
`client_url=$(ETCDADM_MEMBER_INDEX=${next_index} member_client_url)`

But with no endpoint running locally (yet), the etcdctl commands fail... and no local endpoint can be brought up and things continue to fail.

My solution was to set:
```
ETCDCTL_ENDPOINT=https://etcd0:2379,https://etcd1:2379,https://etcd2:2379
```
After that, the etcdctl commands succeed and etcdadm does it's thing and we're off on our merry way.

Now that I had our production etcd up and running, I went over to a non-production cluster and terminated one of the members... and watched it come up without fail.  So, I can't repro it and I don't know where ETCDCTL_ENDPOINT is getting its value since it's not in the environment files or the template.

And that's when @flah00 said, post your results and see if the kube-aws folks have an idea before you go down that rabbit hole.
